code/modules/vending/mail.dm

```diff
@@ -0,0 +1,303 @@
+		user,
+		src,
+		choices,
+		require_near = !HAS_SILICON_ACCESS(user),
+		autopick_single_option = FALSE,
+	)
+	if (!choice)
+		return
+	switch (choice)
+		if ("Eject")
+			pick_mail(user)
+		if ("Dump")
+			playsound(src, 'sound/machines/buzz/buzz-sigh.ogg', 20, TRUE)
+			to_chat(user, span_notice("[src] dumps [length(mail_list)] envelope\s on the floor."))
+			dump_all_mail()
+		if ("Sort")
+			sort_mail(user)
+
+/obj/machinery/mailsorter/proc/sort_mail(mob/user)
+	var/sorting_dept = tgui_input_list(user, "Choose the department to sort mail for","Mail Sorting", sorting_departments)
+	if (!sorting_dept)
+		return
+	currentstate = STATE_SORTING
+	update_appearance(UPDATE_OVERLAYS)
+	playsound(src, 'sound/machines/mail_sort.ogg', 20, TRUE)
+	addtimer(CALLBACK(src, PROC_REF(continue_sort), user, sorting_dept), 5 SECONDS)
+
+
+/obj/machinery/mailsorter/proc/continue_sort(mob/user, sorting_dept)
+	var/list/sorted_mail = list()
+	var/total_to_sort = length(mail_list)
+	var/sorted = 0
+	var/unable_to_sort = 0
+
+	for (var/obj/item/mail/some_mail in mail_list)
+		if (!some_mail.recipient_ref)
+			unable_to_sort ++
+			continue
+		var/datum/mind/some_recipient = some_mail.recipient_ref.resolve()
+		if (some_recipient)
+			var/datum/job/recipient_job = some_recipient.assigned_role
+			var/datum/job_department/primary_department = recipient_job.departments_list?[1]
+			var/datum/job_department/main_department = primary_department.department_name
+			if (main_department == sorting_dept)
+				sorted_mail.Add(some_mail)
+				sorted ++
+		else
+			unable_to_sort ++
+	if (length(sorted_mail) == 0)
+		currentstate = STATE_NO
+		update_appearance(UPDATE_OVERLAYS)
+		playsound(src, 'sound/machines/buzz/buzz-sigh.ogg', 20, TRUE)
+		say("No mail for the following department: [sorting_dept].")
+	else
+		currentstate = STATE_YES
+		update_appearance(UPDATE_OVERLAYS)
+		say("[sorted] envelope\s sorted successfully.")
+		playsound(src, 'sound/machines/ping.ogg', 20, TRUE)
+		to_chat(user, span_notice("[src] ejects [length(sorted_mail)] envelope\s."))
+		var/turf/unload_turf = get_unload_turf()
+		for (var/obj/item/mail/mail_in_list in sorted_mail)
+			mail_in_list.forceMove(unload_turf)
+			sorted_mail -= mail_in_list
+			mail_list -= mail_in_list
+	addtimer(CALLBACK(src, PROC_REF(check_sorted), unable_to_sort, total_to_sort), 1 SECONDS)
+
+/obj/machinery/mailsorter/proc/check_sorted(mob/user, unable_to_sort, total_to_sort)
+	if (unable_to_sort > 0)
+		playsound(src, 'sound/machines/buzz/buzz-sigh.ogg', 20, TRUE)
+		say("Couldn't sort [unable_to_sort] envelope\s.")
+	else
+		playsound(src, 'sound/machines/ping.ogg', 20, TRUE)
+		say("[total_to_sort] envelope\s processed.")
+	addtimer(CALLBACK(src, PROC_REF(update_state_after_sorting)), 1 SECONDS)
+
+/obj/machinery/mailsorter/proc/update_state_after_sorting()
+	currentstate = STATE_IDLE
+	update_appearance(UPDATE_OVERLAYS)
+
+/obj/machinery/mailsorter/attackby(obj/item/thingy, mob/user, params)
```

Comment:
I think this can use interact_with_atom too

Reply:
Is this generally better to use? `attackby` is used by most of the stuff I've seen, what's the difference?